server/sttp-stub-server/src/main/scala/sttp/tapir/server/stub/TapirStubInterpreter.scala

```diff
@@ -0,0 +1,94 @@
+    new SttpBackend[F, R] {
+      override def send[T, P](request: Request[T, P]): F[Response[T]] = {
+        // SttpBackendStub is used to send request since it adjusts response to the shape described by request
+        new SttpBackendStub[F, P](monad, { case _ => StubServerInterpreter(request, endpoints, interceptors) }, None).send(request)
+      }
+
+      override def close(): F[Unit] = monad.unit(())
+      override def responseMonad: MonadError[F] = monad
+    }
+  }
+
+  class TapirEndpointStub[I, E, O](ep: Endpoint[_, I, E, O, _]) {
+    def returnSuccess(response: O): TapirStubInterpreter[F, R, OPTIONS] =
+      append(publicEndpoint(logic = _ => _ => (Right(response): Either[E, O]).unit))
+
+    def returnError(errorResponse: E): TapirStubInterpreter[F, R, OPTIONS] =
+      append(publicEndpoint(logic = _ => _ => (Left(errorResponse): Either[E, O]).unit))
+
+    def throwException(ex: Throwable): TapirStubInterpreter[F, R, OPTIONS] =
+      append(publicEndpoint(logic = _ => _ => throw ex))
+
+    private def publicEndpoint(logic: MonadError[F] => I => F[Either[E, O]]): ServerEndpoint[R, F] =
+      ServerEndpoint.public[I, E, O, R, F](
+        sttp.tapir.endpoint.in(ep.input).out(ep.output).errorOut(ep.errorOutput).asInstanceOf[Endpoint[Unit, I, E, O, R]],
+        logic
+      )
+  }
+
+  class TapirServerEndpointStub[A, U, I, E, O](sep: ServerEndpoint.Full[A, U, I, E, O, R, F]) {
+    def returnSuccess(response: O, runSecurityLogic: Boolean = true): TapirStubInterpreter[F, R, OPTIONS] =
+      if (runSecurityLogic) {
+        append(
+          ServerEndpoint[A, U, I, E, O, R, F](
+            sep.endpoint,
+            sep.securityLogic,
+            _ => _ => _ => (Right(response): Either[E, O]).unit
+          )
+        )
+      } else new TapirEndpointStub(sep.endpoint).returnSuccess(response)
+
+    def returnError(errorResponse: E, runAuthLogic: Boolean = true): TapirStubInterpreter[F, R, OPTIONS] =
+      if (runAuthLogic) {
+        append(
+          ServerEndpoint[A, U, I, E, O, R, F](
+            sep.endpoint,
+            sep.securityLogic,
+            _ => _ => _ => (Left(errorResponse): Either[E, O]).unit
+          )
+        )
+      } else new TapirEndpointStub(sep.endpoint).returnError(errorResponse)
+
+    def throwException(ex: Throwable): TapirStubInterpreter[F, R, OPTIONS] = new TapirEndpointStub(sep.endpoint).throwException(ex)
+
+    def runLogic(): TapirStubInterpreter[F, R, OPTIONS] = append(sep)
+  }
+
+  private def append(sep: ServerEndpoint[R, F]) = new TapirStubInterpreter[F, R, OPTIONS](endpoints :+ sep, interceptors)
+}
+
+object TapirStubInterpreter {
+  def apply[F[_], R, O](options: CustomInterceptors[F, O], monad: MonadError[F]): TapirStubInterpreter[F, R, O] =
+    new TapirStubInterpreter[F, R, O](endpoints = List.empty, options.interceptors)(monad)
+
+  def apply[F[_], R](interceptors: List[Interceptor[F]], monad: MonadError[F]): TapirStubInterpreter[F, R, Any] =
```

Comment:
maybe we can avoid having to pass the `MonadError` parameter? it's mostly hidden from users

Reply:
With the current implementation I create an instance of `SttpBackend`:
```
  /** Returns `SttpBackend` which runs `ServerInterpreter` on each request */
  def backend(): SttpBackend[F, R] = {
    new SttpBackend[F, R] {
      override def send[T, P](request: Request[T, P]): F[Response[T]] = {
        // SttpBackendStub is used to send request since it adjusts response to the shape described by request
        new SttpBackendStub[F, P](monad, { case _ => StubServerInterpreter(request, endpoints, interceptors) }, None).send(request)
      }

      override def close(): F[Unit] = monad.unit(())
      override def responseMonad: MonadError[F] = monad
    }
  }
```
Initially I implemented `TapirStubInterpreter` so that it needs an instance of `SttpBackendStub` (which has a `MonadError` inside) and `backend` method was as follows:
```
  /** Returns `SttpBackendStub` which runs `ServerInterpreter` on each request */
  def backend(): SttpBackendStub[F, R] =
    stub.whenAnyRequest.thenRespondF(req => TestServerInterpreter[F, R](req, endpoints, interceptors))
```
then I thought it's not that good since it allows user to define more `when/then's` on the returned stub and that may "corrupt" the actual intent of running each request against interpreter

Reply:
hm I see, we need to monad. Though I think your initial design was better, it's well-documented how to create a `SttpBackendStub`, as opposed to "where do I get a MonadError from". If we say that the interpreter adds behavior to the stub, we should be fine, if users want to create corrupt stubs we can't stop them ;-)

Reply:
okay, so I will revert to the previous version